cloud/edgecontroller/pkg/devicecontroller/controller/downstream.go

```diff
@@ -0,0 +1,742 @@
+type CacheDeviceModel struct {
+	metav1.ObjectMeta
+	Spec v1alpha1.DeviceModelSpec
+}
+
+// DownstreamController watch kubernetes api server and send change to edge
+type DownstreamController struct {
+	kubeClient   *kubernetes.Clientset
+	messageLayer messagelayer.MessageLayer
+
+	deviceManager *manager.DeviceManager
+	deviceStop    chan struct{}
+
+	deviceModelManager *manager.DeviceModelManager
+	deviceModelStop    chan struct{}
+
+	configMapManager *manager.ConfigMapManager
+
+	crdClient *rest.RESTClient
+}
+
+// syncDeviceModel is used to get events from informer
+func (dc *DownstreamController) syncDeviceModel(stop chan struct{}) {
+	running := true
+	for running {
+		select {
+		case e := <-dc.deviceModelManager.Events():
+			deviceModel, ok := e.Object.(*v1alpha1.DeviceModel)
+			if !ok {
+				log.LOGGER.Warnf("object type: %T unsupported", deviceModel)
+				continue
+			}
+			switch e.Type {
+			case watch.Added:
+				dc.deviceModelAdded(deviceModel)
+			case watch.Deleted:
+				dc.deviceModelDeleted(deviceModel)
+			case watch.Modified:
+				dc.deviceModelUpdated(deviceModel)
+			default:
+				log.LOGGER.Warnf("deviceModel event type: %s unsupported", e.Type)
+			}
+		case <-stop:
+			log.LOGGER.Infof("stop syncDeviceModel")
+			running = false
+		}
+	}
+}
+
+func (dc *DownstreamController) deviceModelAdded(deviceModel *v1alpha1.DeviceModel) {
+	// nothing to do when deviceModel added, only add in map
+	dc.deviceModelManager.DeviceModel.Store(deviceModel.Name, &CacheDeviceModel{ObjectMeta: deviceModel.ObjectMeta, Spec: deviceModel.Spec})
+}
+
+func isDeviceModelUpdated(old *CacheDeviceModel, new *v1alpha1.DeviceModel) bool {
+	// does not care fields
+	old.ObjectMeta.ResourceVersion = new.ObjectMeta.ResourceVersion
+	old.ObjectMeta.Generation = new.ObjectMeta.Generation
+
+	// return true if ObjectMeta or Spec or Status changed, else false
+	return !reflect.DeepEqual(old.ObjectMeta, new.ObjectMeta) || !reflect.DeepEqual(old.Spec, new.Spec)
+}
+
+func (dc *DownstreamController) deviceModelUpdated(deviceModel *v1alpha1.DeviceModel) {
+	value, ok := dc.deviceModelManager.DeviceModel.Load(deviceModel.Name)
+	dc.deviceModelManager.DeviceModel.Store(deviceModel.Name, &CacheDeviceModel{ObjectMeta: deviceModel.ObjectMeta, Spec: deviceModel.Spec})
+	if ok {
+		cachedDeviceModel := value.(*CacheDeviceModel)
+		if isDeviceModelUpdated(cachedDeviceModel, deviceModel) {
+			dc.updateAllConfigMaps(deviceModel)
+		}
+	} else {
+		dc.deviceModelAdded(deviceModel)
+	}
+}
+
+func (dc *DownstreamController) updateAllConfigMaps(deviceModel *v1alpha1.DeviceModel) {
+	//TODO: add logic to update all config maps, How to manage if a property is deleted but a device is referring that property. Need to come up with a design.
+}
+
+func (dc *DownstreamController) deviceModelDeleted(deviceModel *v1alpha1.DeviceModel) {
+	// TODO: Need to use finalizer like method to delete all devices referring to this model. Need to come up with a design.
```

Comment:
maybe it is other way around, like only delete the model if no devices are referring to it ?

Reply:
I think if a deviceModel is deleted in apiserver, all devices referring to it should be deleted or we do not allow to delete the model. Model can be present without a device as model is independent of device, but device is dependent on model